### PR TITLE
test(scorecard): scorecard test for Cryostat CR configuration changes

### PR DESCRIPTION
**Pull request by mwangggg**
# Welcome to Cryostat! 👋
## Before contributing, make sure you have:
* [x] Read the [contributing guidelines](https://github.com/cryostatio/cryostat/blob/main/CONTRIBUTING.md)
* [x] Linked a relevant issue which this PR resolves
* [x] Linked any other relevant issues, PR's, or documentation, if any
* [x] Resolved all conflicts, if any
* [x] Rebased your branch PR on top of the latest upstream `main` branch
* [x] Attached at least one of the following labels to the PR: `[chore, ci, docs, feat, fix, test]`
* [x] [Signed all commits](https://docs.github.com/en/authentication/managing-commit-signature-verification/signing-commits): `git commit -S -m "YOUR_COMMIT_MESSAGE"`
_______________________________________________

Fixes: #503 

## Description of the change:
Updates the Cryostat CR from default emptyDir storage to PVC storage and checks if the resulting redeployment is as expected.

## How to manually test:
regular steps to check scorecard tests

### CODE REVIEW

**Comment by mwangggg** [comment]
https://github.com/cryostatio/cryostat-operator/actions/runs/8073777424

File: hack/custom.config.yaml.in [hunk 1]
```diff
@@ -31,3 +31,13 @@
     labels:
       suite: cryostat
       test: cryostat-recording
+- op:add
```
Comment (ebaron) [inline hack/custom.config.yaml.in:34]: I think this needs a space between `op:` and `add`, and then a `make bundle`. I was getting an error when running `make bundle`.

**Review by ebaron — COMMENTED** [review]
Hey Ming, nice work! Just a couple of small things.

File: internal/test/scorecard/common_utils.go [hunk 2]
```diff
@@ -383,6 +383,88 @@ func waitTillCryostatReady(base *url.URL, resources *TestResources) error {
...
+				if condition.Type == appsv1.DeploymentProgressing && condition.Status == corev1.ConditionFalse && condition.Reason == "ProgressDeadlineExceeded" {
+					return false, fmt.Errorf("deployment %s exceeded its progress deadline", deploy.Name) // Don't Retry
+				}
+			}
+			if deploy.Spec.Replicas != nil && deploy.Status.UpdatedReplicas < *deploy.Spec.Replicas {
+				r.Log += fmt.Sprintf("Waiting for deployment %s rollout to finish: %d out of %d new replicas have been updated... \n", deploy.Name, deploy.Status.UpdatedReplicas, *deploy.Spec.Replicas)
+				return false, nil
+			}
+			if deploy.Status.Replicas > deploy.Status.UpdatedReplicas {
+				r.Log += fmt.Sprintf("Waiting for deployment %s rollout to finish: %d old replicas are pending termination.. \n", deploy.Name, deploy.Status.Replicas-deploy.Status.UpdatedReplicas)
```
Comment (ebaron) [inline internal/test/scorecard/common_utils.go:432]: There's a missing `.` at the end of this log message

File: internal/test/scorecard/tests.go [hunk 3]
```diff
@@ -124,6 +183,7 @@ func CryostatRecordingTest(bundle *apimanifests.Bundle, namespace string, openSh
 		Password:        string(secret.Data["CRYOSTAT_RJMX_PASS"]),
 		MatchExpression: fmt.Sprintf("target.alias==\"%s\"", target.Alias),
 	}
+	defer cleanupCryostat(r, tr.Client, CryostatCRTestName, namespace)
```
Comment (tthvo) [inline internal/test/scorecard/tests.go:186]: I think we can remove it? On line 149, we defer cleanup

https://github.com/cryostatio/cryostat-operator/blob/cfcbfc74fab65b78bf5fb63a8c5abe658d65fe6a/internal/test/scorecard/tests.go#L90

File: internal/test/scorecard/tests.go [hunk 4]
```diff
@@ -72,6 +76,61 @@ func CryostatCRTest(bundle *apimanifests.Bundle, namespace string, openShiftCert
...
+	cr := newCryostatCR(CryostatConfigChangeTestName, namespace, !tr.OpenShift)
+	cr.Spec.StorageOptions = &operatorv1beta1.StorageConfiguration{
+		EmptyDir: &operatorv1beta1.EmptyDirConfig{
+			Enabled: true,
+		},
+	}
+
+	_, err = createAndWaitTillCryostatAvailable(cr, tr)
+	if err != nil {
+		return fail(*r, fmt.Sprintf("%s test failed: %s", CryostatConfigChangeTestName, err.Error()))
```
Comment (tthvo) [inline internal/test/scorecard/tests.go:98]: Maybe, let's use a similar log message to other tests?

https://github.com/cryostatio/cryostat-operator/blob/cfcbfc74fab65b78bf5fb63a8c5abe658d65fe6a/internal/test/scorecard/tests.go#L88

File: internal/test/scorecard/tests.go [hunk 5]
```diff
@@ -72,6 +76,61 @@ func CryostatCRTest(bundle *apimanifests.Bundle, namespace string, openShiftCert
...
+		},
+	}
+
+	// Wait for redeployment of Cryostat CR
+	err = updateAndWaitTillCryostatAvailable(cr, tr)
+	if err != nil {
+		return fail(*r, fmt.Sprintf("Cryostat redeployment did not become available: %s", err.Error()))
+	}
+	r.Log += "Cryostat deployment has successfully updated to pvc"
+	defer cleanupCryostat(r, tr.Client, CryostatConfigChangeTestName, namespace)
```
Comment (tthvo) [inline internal/test/scorecard/tests.go:129]: Would it make sense to call defer after 

```bash
_, err = createAndWaitTillCryostatAvailable(cr, tr)
```

That way, if any errors occur, the CR is cleaned up regardless?

File: internal/test/scorecard/tests.go [hunk 6]
```diff
@@ -72,6 +76,61 @@ func CryostatCRTest(bundle *apimanifests.Bundle, namespace string, openShiftCert
...
+			},
+		},
+	}
+
+	// Wait for redeployment of Cryostat CR
+	err = updateAndWaitTillCryostatAvailable(cr, tr)
+	if err != nil {
+		return fail(*r, fmt.Sprintf("Cryostat redeployment did not become available: %s", err.Error()))
+	}
+	r.Log += "Cryostat deployment has successfully updated to pvc"
```
Comment (tthvo) [inline internal/test/scorecard/tests.go:128]: Maybe like: `Cryostat deployment has successfully updated and rolled out`?

File: internal/test/scorecard/common_utils.go [hunk 7]
```diff
@@ -383,6 +383,88 @@ func waitTillCryostatReady(base *url.URL, resources *TestResources) error {
...
+		if len(cr.Status.ApplicationURL) > 0 {
+			return true, nil
+		}
+		r.Log += "application URL is not yet available\n"
+		return false, nil
+	})
+	if err != nil {
+		return fmt.Errorf("application URL not found in CR: %s", err.Error())
+	}
+	r.Log += fmt.Sprintf("application is available at %s\n", cr.Status.ApplicationURL)
```
Comment (tthvo) [inline internal/test/scorecard/common_utils.go:463]: At the stage, the applicationURL is always available on the CR status right? How about checking cryostat's health endpoint instead?

https://github.com/cryostatio/cryostat-operator/blob/cfcbfc74fab65b78bf5fb63a8c5abe658d65fe6a/internal/test/scorecard/common_utils.go#L339

File: internal/test/scorecard/clients.go [hunk 8]
```diff
@@ -158,9 +158,9 @@ func create[r runtime.Object](ctx context.Context, c rest.Interface, res string,
 	return result, err
 }
 
-func update[r runtime.Object](ctx context.Context, c rest.Interface, res string, ns string, obj r, result r) (r, error) {
+func update[r runtime.Object](ctx context.Context, c rest.Interface, res string, ns string, obj r, result r, name string) (r, error) {
 	err := c.Put().
-		Namespace(ns).Resource(res).
+		Namespace(ns).Resource(res).Name(name).
```
Comment (tthvo) [inline internal/test/scorecard/clients.go:163]: Just curious here: The body is populated with the object, which should have the object name in metadata already right? Or am I misunderstanding...?

**Review by tthvo — COMMENTED** [review]
Just some thoughts I had when Ming and I was discussing this PR. I hope it helps in a way :D

File: internal/test/scorecard/clients.go (hunk 8 above)
Comment (mwangggg) [reply internal/test/scorecard/clients.go:163]: When I didn't send the name directly into the request I would get an error about a missing name?

Comment (tthvo) [reply internal/test/scorecard/clients.go:163]: Ahh i see thanks! I looked into deployment client and they do the same :))

**Review by ebaron — APPROVED** [review]
Looks good to me, thanks @mwangggg!

**Comment by ebaron** [comment]
@Mergifyio backport cryostat3

**Pull request merged**